### PR TITLE
configurable devEnabled in React.gradle

### PR DESCRIPTION
**Pull request by vonovak**
This PR adds support for configurable devEnabled option when building an android app. This is currently hardcoded.

The reason for making this configurable is this: I have an app that uses code-push and 3 buildConfigs. I want to have a debugging version which has `devEnabled = true`, then a staging version which has `devEnabled = false` (this version of the app is used internally for testing and should behave just like a release version of the app, and when the tests succeed the changes are promoted into the release version to the users out there, using code-push). The last version is a standard release version with `devEnabled = false`.
 
Currently, `devEnabled` is hardwired like this: `!targetName.toLowerCase().contains("release")` so by default my `staging` buildConfig will have `devEnabled = true` but I'd like it to be false.

With this PR it'd be possible to configure this as follows, while not breaking the current behavior.
`'devDisabledIn${productFlavor}${buildType}'`
`'devDisabledIn${buildType}'`

eg:
```
project.ext.react = [
        devDisabledInStaging: true
]
```

### Testing

 When adding a println statement to `react.gradle` I get the following output for the following config (for full and demo build flavors):

```
project.ext.react = [
        devDisabledInDemoStaging: true,
        devDisabledInDemoDebug: false
]
```

```
targetName DemoDebug, buildTypeName Debug, devEnabled: true
targetName DemoRelease, buildTypeName Release, devEnabled: false
targetName DemoStaging, buildTypeName Staging, devEnabled: false
targetName FullDebug, buildTypeName Debug, devEnabled: true
targetName FullRelease, buildTypeName Release, devEnabled: false
targetName FullStaging, buildTypeName Staging, devEnabled: true

```

so setting to both false and true is picked up.

Let me know if you're interested in this and if yes, I'd add a documentation for this to the [build.gradle template](https://github.com/facebook/react-native/blob/a477aec10d29b4651944cb4292daf06dfa953ea3/local-cli/templates/HelloWorld/android/app/build.gradle#L25)

note this does not allow `devDisabledInRelease: false` which is something you'd never want

### CODE REVIEW

**Comment by hramos** [comment]
This seems useful. One minor nit, why did you pick devDisabled as the config name instead of devEnabled?

**Comment by vonovak** [comment]
@hramos good question.
This is the current implementation: `def devEnabled = !targetName.toLowerCase().contains("release")` and I know didn't want to change this part since I thought it might (?) break something.

Originally, I wrote
```
def devEnabled
if (config."devEnabledIn${targetName}" != null) {
    devEnabled = config."devEnabledIn${targetName}"
} else if(config."devEnabledIn${buildTypeName.capitalize()}" != null) {
    devEnabled = config."devEnabledIn${buildTypeName.capitalize()}"
} else {
    devEnabled = !targetName.toLowerCase().contains("release")
}
```

but the code isn't very pretty and it allows to say `devEnabledInInRelease: true` which isn't something that should be allowed imo, and it would deviate from the current behavior.

Now I don't quite remember why I did it with `devDisabled`, but looking at the code, I could as well say 

```
def devEnabled = !(!config."devEnabledIn${targetName}"
    || !config."devEnabledIn${buildTypeName.capitalize()}"
    || targetName.toLowerCase().contains("release"))
```

which is the equivalent of what is in the PR, just with two extra `!`. I think I like this naming slightly better. I'll change it to this and also update the [template](https://github.com/facebook/react-native/blob/a477aec10d29b4651944cb4292daf06dfa953ea3/local-cli/templates/HelloWorld/android/app/build.gradle#L25) if you agree.

**Comment by hramos** [comment]
No need to change, I'm OK with going with "devDisabled**" here. Go ahead and update the template.

File: react.gradle [hunk 1]
```diff
@@ -76,7 +76,9 @@ gradle.projectsEvaluated {
                 workingDir reactRoot
 
                 // Set up dev mode
-                def devEnabled = !targetName.toLowerCase().contains("release")
+                def devEnabled = !(config."devDisabledIn${targetName}"
+                    || config."devDisabledIn${buildTypeName.capitalize()}"
+                    || targetName.toLowerCase().contains("release"))
```
Comment (AndrewJack) [inline react.gradle]: This is just the same as:
```
def devEnabled = !(config."devDisabledIn${targetName}"
        || targetName.toLowerCase().contains("release"))
```

productFlavorName is an empty string if no flavors are specified, making `targetName` the same as `buildTypeName`
Could you remove line 80?

Comment (vonovak) [inline react.gradle]: yeah, I see now. Will change.

**Comment by chrisknepper** [comment]
Can we please get this merged in? It would really be great to be able to name our `buildTypes` what we want instead of including the word "release" because of a single hardcoded word in `react.gradle`.

**Comment by hramos** [comment]
@chrisknepper I can take a look at this. I missed that @vonovak had updated the PR, please always make sure to tag us as we can easily miss these follow up commits.

**Comment by vonovak** [comment]
@hramos sorry for not tagging you, I waited until the tests would be successful and then I went for a long vacation (btw. for re-running the tests, I'd have to push something to my branch even though I didn't want to modify anything. I used the "update branch" button instead since it's so quick - what is the preferred way to re-run tests if they previously failed?).

not sure why travis is complaining, my PR doesn't seem to be related..?

**Comment by hramos** [comment]
We're still working on importing this internally. There's been some internal test failures that seem unrelated and hoping a retry will let this land.

**Comment by vonovak** [comment]
@hramos this should be good to be merged now

**Comment by maxs15** [comment]
Need this PR so bad, hope it'll be merged soon :crossed_fingers:

**Pull request closed without merging**